### PR TITLE
Hardhat tests

### PR DESCRIPTION
**Pull request by SoulFletcher**

### CODE REVIEW

File: packages/hardhat/contracts/ChainLance.sol [hunk 1]
```diff
@@ -278,7 +278,7 @@ contract ChainLance {
         require(projectList.contains(projectId), "unknown project");
         require(projects[projectId].state == ProjectState.Open, "not open");
         require(projects[projectId].owner == msg.sender, "not owner");
-        projects[projectId].state == ProjectState.Canceled;
+        projects[projectId].state = ProjectState.Canceled; // fix ("=" and not "==")
```
Comment (gladcow) [inline packages/hardhat/contracts/ChainLance.sol]: Remove this comment

File: packages/hardhat/test/ChainLance.ts [hunk 2]
```diff
@@ -13,15 +13,240 @@ describe("ChainLance", function () {
...
+    const receiptSubmitWork = await txSubmitWork.wait();
+    expect(receiptSubmitWork?.status).is.equal(1, "Work submition failed");
+    ///////////////////////////////////////////////////////////////////
+    // Accept Work
+    ///////////////////////////////////////////////////////////////////
+    const txAcceptWork = await chainLance.connect(employer).acceptWork(projectId);
+    const receiptAcceptWork = await txAcceptWork.wait();
+    expect(receiptAcceptWork?.status).is.equal(1, "Work acception failed");
+    if (receiptAcceptWork) {
+      expect(await ethers.provider.getBalance(worker.address)).is.greaterThanOrEqual(
```
Comment (gladcow) [inline packages/hardhat/test/ChainLance.ts]: Worker can have some initial balance in some setups. You should request worker balance before accepting the work (it is equal to zero in our case) and subtract it from the balance after accepting to compare it with project price.

File: packages/hardhat/test/ChainLance.ts [hunk 3]
```diff
@@ -13,15 +13,240 @@ describe("ChainLance", function () {
...
+    const txRejectWork = await chainLance.connect(employer).rejectWork(projectId);
+    const receiptRejectWork = await txRejectWork.wait();
+    expect(receiptRejectWork?.status).is.equal(1, "work rejection failed");
+
+    ///////////////////////////////////////////////////////////////////
+    // Cancel work
+    ///////////////////////////////////////////////////////////////////
+
+    await ethers.provider.send("evm_increaseTime", [timeSpanToCancel + 1]); // Changing block timestamp
+    await ethers.provider.send("evm_mine", []); // Mine block
```
Comment (gladcow) [inline packages/hardhat/test/ChainLance.ts]: We don't need this 2 lines here. We are waiting for next block in  `txRejectWork.wait();` If you need wait for more than 1 block,  you can pass value in this function,  for example `txRejectWork.wait(2);` But I don't understand why we need to wait for block at all,  any number of operations on project cancel should work inside 1 block.

File: packages/hardhat/test/ChainLance.ts [hunk 4]
```diff
@@ -13,15 +13,240 @@ describe("ChainLance", function () {
...
+  // We define a fixture to reuse the same setup in every test.
+
+  let chainLance: ChainLance;
+  before(async () => {
+    // const [owner] = await ethers.getSigners();
+    const chainLanceFactory = await ethers.getContractFactory("ChainLance");
+    chainLance = (await chainLanceFactory.deploy()) as ChainLance;
+    await chainLance.waitForDeployment();
+
+    const [employer1, employer2, employer3, worker1, worker2, worker3] = await ethers.getSigners(); // can i do that?
```
Comment (gladcow) [inline packages/hardhat/test/ChainLance.ts]: Yes. REmove this comment. You can leave comments like this one in PR draft instead.

File: packages/hardhat/test/ChainLance.ts [hunk 5]
```diff
@@ -13,15 +13,240 @@ describe("ChainLance", function () {
...
+
+    const bidId7 = "bidid7";
+    await chainLance.connect(worker3).bidProject(canceledProjectId, bidId7, canceledProjectPrice, 1);
+
+    await chainLance.connect(employer2).cancelProject(canceledProjectId);
+  });
+
+  it("Should list projects in different states", async function () {
+    const openProjects = await chainLance.listProjectsWithState(0);
+    expect(openProjects).to.include("id42");
```
Comment (gladcow) [inline packages/hardhat/test/ChainLance.ts:219]: Check the number of elements in all call's results in this case too

File: packages/hardhat/test/ChainLance.ts [hunk 6]
```diff
@@ -13,15 +13,240 @@ describe("ChainLance", function () {
...
+    const completedProjects = await chainLance.listProjectsWithState(3);
+    expect(completedProjects).to.include("id5");
+
+    const canceledProjects = await chainLance.listProjectsWithState(4);
+    expect(canceledProjects).to.include("id6");
+  });
+});
+
+describe("ChainLance: check access rights", function () {
+  // We define a fixture to reuse the same setup in every test.
```
Comment (gladcow) [inline packages/hardhat/test/ChainLance.ts]: Remove this comment

File: packages/hardhat/test/ChainLance.ts [hunk 7]
```diff
@@ -13,15 +13,240 @@ describe("ChainLance", function () {
...
+
+  let chainLance: ChainLance;
+  before(async () => {
+    // const [owner] = await ethers.getSigners();
```
Comment (gladcow) [inline packages/hardhat/test/ChainLance.ts]: remove this line

File: packages/hardhat/test/ChainLance.ts [hunk 8]
```diff
@@ -13,15 +13,240 @@ describe("ChainLance", function () {
...
+    const chainLanceFactory = await ethers.getContractFactory("ChainLance");
+    chainLance = (await chainLanceFactory.deploy()) as ChainLance;
+    await chainLance.waitForDeployment();
+    // TODO: create some projects in different states and several bids on them to check different
+    // they can be used in all "it" sections
+  });
+
+  it("Should allow to change project state to employer only", async function () {
+    // TODO
```
Comment (gladcow) [inline packages/hardhat/test/ChainLance.ts]: implement this case

File: packages/hardhat/test/ChainLance.ts [hunk 9]
```diff
@@ -13,15 +13,240 @@ describe("ChainLance", function () {
...
+  });
+
+  it("Should allow to change work state to worker only", async function () {
+    // TODO
```
Comment (gladcow) [inline packages/hardhat/test/ChainLance.ts]: implement this case

**Review by gladcow — CHANGES_REQUESTED** [review]

File: packages/hardhat/test/ChainLance.ts (hunk 3 above)
Comment (SoulFletcher) [reply packages/hardhat/test/ChainLance.ts]: txRejectWork.wait(timeSpanToCancel + 1); results in error: "Error: Timeout of 40000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves"

File: packages/hardhat/test/ChainLance.ts (hunk 9 above)
Comment (SoulFletcher) [reply packages/hardhat/test/ChainLance.ts]: Can't use vals from "before" in "It"

File: packages/hardhat/test/ChainLance.ts (hunk 3 above)
Comment (gladcow) [reply packages/hardhat/test/ChainLance.ts]: Let's set `timeSpanToCancel` to 1 for cancel work test

File: packages/hardhat/test/ChainLance.ts [hunk 10]
```diff
@@ -13,15 +13,330 @@ describe("ChainLance", function () {
...
+      .createProject(oneBidProjectId, oneBidProjectPrice, 100);
+    await txCreateOneBidProject.wait();
+
+    const bidId1 = "bidid1";
+    await chainLance.connect(worker1).bidProject(oneBidProjectId, bidId1, oneBidProjectPrice, 1);
+
+    const txAcceptBid1 = await chainLance
+      .connect(employer2)
+      .acceptBid(oneBidProjectId, bidId1, { value: oneBidProjectPrice });
+    await txAcceptBid1.wait();
```
Comment (gladcow) [inline packages/hardhat/test/ChainLance.ts:279]: We  should  check contract call is failed here. Get receipt for this transaction (`const receipt = await txAcceptBid1.wait();`) and check receipt.state is 2 (`expect(receipt.state).to.be equal(2, "contract call should fail")`)

File: packages/hardhat/test/ChainLance.ts [hunk 11]
```diff
@@ -13,15 +13,330 @@ describe("ChainLance", function () {
...
+
+    const bidId3 = "bidid3";
+    await chainLance.connect(worker3).bidProject(twoBidsProjectId, bidId3, twoBidsProjectPrice, 1);
+
+    const txAcceptBid2 = await chainLance
+      .connect(employer2)
+      .acceptBid(twoBidsProjectId, bidId2, { value: twoBidsProjectPrice });
+    await txAcceptBid2.wait();
+
+    const txSubmitWork2 = await chainLance.connect(worker2).submitWork(twoBidsProjectId);
```
Comment (gladcow) [inline packages/hardhat/test/ChainLance.ts:298]: we should check other worker or employer can't submit work or cancel bid

File: packages/hardhat/test/ChainLance.ts (hunk 9 above)
Comment (gladcow) [reply packages/hardhat/test/ChainLance.ts]: this vals should be declared as consts or var outside the `before` section

**Review by gladcow — CHANGES_REQUESTED** [review]

**Review by gladcow — APPROVED** [review]

**Pull request merged**